[GTK][WPE] Add webkit://gpu/stdout to dump the info from webkit://gpu to stdout

clopez opened the pull request:
#### a9ad2809c62208af544f47187fc8e4767f4bb1a0
<pre>
[GTK][WPE] Add webkit://gpu/stdout to dump the info from webkit://gpu to stdout
<a href="https://bugs.webkit.org/show_bug.cgi?id=277045">https://bugs.webkit.org/show_bug.cgi?id=277045</a>

Reviewed by Carlos Garcia Campos and Nikolas Zimmermann.

This helps getting the info about the GPU on stdout when working with
remote/headless/virtualized screens where sending input events or
taking screenshots is complicated.

* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::prettyPrintJSON):
(WebKit::WebKitProtocolHandler::handleGPU):

Canonical link: <a href="https://commits.webkit.org/282835@main">https://commits.webkit.org/282835@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f07fbea53232ddb06246f60c9fd8d0b6fdc45ef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64440 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43804 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17036 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68461 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15047 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51521 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15327 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51850 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10378 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67508 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40498 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55761 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32469 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37167 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13921 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/13468 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70162 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8387 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/12979 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59174 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8421 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55850 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59341 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6932 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9763 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39618 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/40696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/41879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40439 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann reviewed (approved):
That is indeed helpful!

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
```diff
@@ -620,7 +620,8 @@ void WebKitProtocolHandler::handleGPU(WebKitURISchemeRequest* request)
 
-    g_string_append(html, "</head><body>");
+    // dump to JS console at startup to make easier getting this info on remote/headless/virtual displays
+    g_string_append(html, "</head><body onload=\"sendToConsole()\">");
```
Do we really want to do this unconditionally? I think we can do this only when requested using the request path, and we could add more actions, for example

* webkit://gpu/json/console: send the json to the console like this patch does.
* webkit://gpu/json/stdout: prints the json in stdout
* webkit://gpu/json/clipboard: writes the json in the clipboard

These options are useful not only for embedded but also when debugging rendering issues where nothing is rendered.

nikolaszimmermann replied:
I would vote to keep this simple, the dumping does not hurt performance and we have less things to remember/document.

I'm not opposed to it though...

nikolaszimmermann replied:
Another point of view:

Dumping to console is enough, we already have a flag to dump console to stdout, and command line tools to copy to clipboard.

carlosgcampos replied:
It's not because of performance, in the gpu page info performance doesn't matter. But what's the point of having a button to send the json to the console if we have already sent it unconditionally on load? Having the info in stout directly is useful too, with this patch we need to enable the setting to send console messages to stout too, which I don't think it's simpler than just typing `webkit://gpu/json/stdout`

carlosgcampos replied:
> Another point of view:
> 
> Dumping to console is enough, we already have a flag to dump console to stdout, and command line tools to copy to clipboard.

what flag do you mean? and what command line to copy to clipboard?

nikolaszimmermann replied:
We have a feature flag, that we can pass as command line argument to dump JS console to stdout - I was referring to that one.

I did not think of the button, in my opinion, it can be removed.

There is no WebKit specific tool to copy to clipboard but e.g. wl-copy.

carlosgcampos replied:
That command line option is present in cog and MiniBrowser, not in all apps using WPE, loading `webkit://gpu/json/stdout `would always be available not matter what browser app is used. Copying to clipboard is specific to GTK port, so `webkit://gpu/json/clipboard` would also be specific to GTK. It has happened many times in gtk port that people reporting nothing is rendered load `webkit://gpu` and click where the button is supposed to be to get the result in the clipboard. In those cases it would be a lot easier to just load a different url, instead of having to enable a setting and reload

clopez replied:
> In those cases it would be a lot easier to just load a different url, instead of having to enable a setting and reload

But enabling the setting is also needed when loading the different url, isn't it?

carlosgcampos replied:
Not when using `webkit://gpu/json/stdout`

carlosgcampos replied:
I see the tricky part is formatting the json like stringify does... So, let's keep it this way, but let's remove the button then, since it's always sent to the console. Also, even if existing issue, we can avoid the utf8 conversion made twice by doing:

```
auto infoAsString = jsonObject->toJSONString().utf8();
```

And then using the CString in both `copyAsJSON` and `sendToConsole`. In the case of WPE we don't need to generate the `copyAsJSON` since we don't add the button.

nikolaszimmermann replied:
The absence of the command line option in non-cog/non-MiniBrowser scenarios convinced me, good point @carlosgcampos.

nikolaszimmermann replied:
However also see the other point you raised :-) It's up to @clopez to decide how to move on :-)

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
```diff
@@ -287,8 +289,72 @@ static String renderBufferFormat(WebKitURISchemeRequest* request)
 #endif
 #endif
 
+String prettyPrintJSON(const String& jsonString)
```
```suggestion
static String prettyPrintJSON(const String& jsonString)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
```diff
@@ -287,8 +289,72 @@ static String renderBufferFormat(WebKitURISchemeRequest* request)
+    bool inQuotes = false;
+    bool escape = false;
+    // 4 spaces per identation level
+    auto s_identSpaceLevel = String("    "_s);
```
```suggestion
    constexpr auto identSpaceLevel = "    "_s;
```

clopez committed:
[GTK][WPE] Add webkit://gpu/stdout to dump the info from webkit://gpu to stdout
https://bugs.webkit.org/show_bug.cgi?id=277045

Reviewed by Carlos Garcia Campos and Nikolas Zimmermann.

This helps getting the info about the GPU on stdout when working with
remote/headless/virtualized screens where sending input events or
taking screenshots is complicated.

* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::prettyPrintJSON):
(WebKit::WebKitProtocolHandler::handleGPU):

Canonical link: https://commits.webkit.org/282835@main
